### PR TITLE
React setup and project refactoring

### PR DESCRIPTION
**Pull request by ThalusA**

### CODE REVIEW

**Review by lambdina — APPROVED** [review]

File: package.json [hunk 1]
```diff
@@ -1,32 +1,74 @@
...
-    "ganache": "ganache-cli --deterministic",
-    "migrate": "truffle migrate",
-    "test": "truffle compile && ts-mocha --exit --recursive --extension js,ts",
-    "watch": "yarn test --watch",
-    "remix": "remixd -s $(pwd) --remix-ide https://remix.ethereum.org",
-    "zokrates": "node zokrates/zokrates.js"
+    "start": "yarn --cwd ./src/contracts build && react-app-rewired start",
+    "build": "yarn --cwd ./src/contracts build && react-app-rewired build",
+    "eject": "react-app-rewired eject",
+    "test": "mocha"
```
Comment (Aelerinya) [inline package.json]: Why is the test script only `mocha` now ?

File: src/app/App.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+          <MetaMask />
+        </div>
+        <div className="app-center">
+          <IpfsHandling />
+        </div>
+        <div className="app-right">
+          <Review />
+        </div>
+      </div>
+    </div>
```
Comment (Aelerinya) [inline src/app/App.tsx]: You could maybe use some more semantic elements like `<section>` or `<article>`

File: src/components/ButtonUpload/ButtonUpload.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    async function submitPaper() {
+        if (filename && fileContent) {
+            const ext = filename.split('.').pop() ?? ".txt";
+            await IPFS.files.mkdir("/reactrs-paper");
+            await IPFS.files.write(`/reactrs-paper/${paperField}${ext}`, fileContent, { create: true });
+            const direntry = await IPFS.files.stat("/reactrs-paper");
+            setFileHash(direntry.cid);
+        }
+    };
```
Comment (Aelerinya) [inline src/components/ButtonUpload/ButtonUpload.tsx]: The name of the file is already stored in the `File` object. No need to store it separately.

```suggestion
    const [fileContent, setFileContent] = React.useState<File | null>(null);
    const [fileHash, setFileHash] = React.useState<CID | null>(null);
    const [paperField, setPaperField] = React.useState("");

    function changeField(event: ChangeEvent<HTMLInputElement>) {
        setPaperField(event.target.value);
    }

    function changeFile(event: ChangeEvent<HTMLInputElement>) {
        setFileContent(event.target.files?.[0] ?? null);
    }

    async function submitPaper() {
        if (fileContent) {
            const ext = fileContent.name.split('.').pop() ?? ".txt";
            await IPFS.files.mkdir("/reactrs-paper");
            await IPFS.files.write(`/reactrs-paper/${paperField}${ext}`, fileContent, { create: true });
            const direntry = await IPFS.files.stat("/reactrs-paper");
            setFileHash(direntry.cid);
        }
    };
```

File: src/components/Review/ChangeRequest.tsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+export default function ChangeRequest({changeRequests, setChangeRequests, value, index}: {
+    changeRequests: Array<ChangeRequestElement>,
+    setChangeRequests: React.Dispatch<React.SetStateAction<ChangeRequestElement[]>>,
+    value: ChangeRequestElement,
+    index: number
+}) {
+    function handleChangeChangeRequestComment(event: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>, index: number) {
+        const newArr = [...changeRequests];
+        newArr[index] = {...newArr[index], comment: event.target.value};
```
Comment (Aelerinya) [inline src/components/Review/ChangeRequest.tsx]: I think you can just edit the object directly
```suggestion
        newArr[index].comment: event.target.value;
```

File: src/components/Review/ChangeRequests.tsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
+import React, { MouseEvent } from 'react';
+import "./ChangeRequests.css";
+import ChangeRequest, { ChangeRequestElement } from "./ChangeRequest";
+
+import ClearIcon from '@material-ui/icons/Clear';
+import Button from '@material-ui/core/Button';
+
+export default function ChangeRequests({hash, reviewStatus}: {hash: string, reviewStatus: string}) {
```
Comment (Aelerinya) [inline src/components/Review/ChangeRequests.tsx]: The same `ChangeRequest` and `ChangeRequests` are really confusing. Consider giving them more descriptive names.

File: src/components/User/UserInteract.tsx [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        
+    }
+
+    function handleClickAddField() {
+        setUserInfos({...userInfos, fields: [...userInfos.fields, field]});
+        setField('');
+    }
+
+    return (
+        <Fragment>
```
Comment (Aelerinya) [inline src/components/User/UserInteract.tsx]: Tip for the next time you write some React: using empty html tags is equivalent to using `React.Fragment`. So `<></>` is the same as `<Fragment></Fragment>`.

File: src/utils/Utils.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+export function truncate(str: string, n: number) {
```
Comment (Aelerinya) [inline src/utils/Utils.ts]: Please rename this function to something like `shortenTextWithEllipsis` or something. It is not doing truncation, so it's confusing to call it `truncate`.

**Review by Aelerinya — APPROVED** [review]
Don't forget to resolve the merge conflicts.

File: package.json (hunk 1 above)
Comment (ThalusA) [reply package.json]: It now invoke test script on contract and does it too with ts-mocha for react env.

**Pull request merged**